// chief/helpers/index.js

module.exports = {


	// core helpers
	extend : require('./core/_extend'),

	index : require('./core/_index'),

    access : require('./core/_access'),

	read : require('./core/_read'),

	write : require('./core/_write'),

	erase : require('./core/_erase'),

	glob : require('glob').sync,

	capitalise : require('./core/_capitalise'),

    stopwatch : require('./core/_stopwatch'),


	// filter helpers
	reduceDirectories : require('./filters/_reduce-directories'),

    diffDirectories : require('./filters/_diff-directories'),

    updateDirectories : require('./filters/_update-directories'),

    reduceFiles : require('./filters/_reduce-files'),

    diffFiles : require('./filters/_diff-files'),

    updateFiles : require('./filters/_update-files'),


    // modifier helpers
    setLocal : require('./modifiers/_set-local'),

    setTemplate : require('./modifiers/_set-template'),

    getTemplate : require('./modifiers/_get-template'),


    // message helpers
    message : require('./message'),
    

    // error helpers
    error : require('./error')

};
